File: 200_Number_of_Islands/numIslands.py
class Solution(object):
    def __init__(self):
        # go right, left, up, down
        self.dx = [1, -1, 0, 0]
        self.dy = [0, 0, 1, -1]
        self.y_lenght = 0
        self.x_length = 0
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        self.x_length = len(grid[0])
        self.y_length = len(grid)

        def removeIsland(y, x):
            grid[y][x] = '0'
            for i in xrange(4):
                nextX = x + self.dx[i]
                nextY = y + self.dy[i]

                if 0 <= nextX < self.x_length and \
                   0 <= nextY < self.y_length and \
                   grid[nextY][nextX] == '1':
                       removeIsland(nextY, nextX)
        res = 0
        for y in xrange(self.y_length):
            for x in xrange(self.x_length):
                if grid[y][x] == '1':
                    res += 1
                    removeIsland(y, x)
                else:
                    continue
        return res

sol = Solution()
grid = [['1', '1', '1', '1', '0'], ['1', '1', '0', '1', '0'], ['1', '1', '0', '0', '0'], ['0', '0', '0', '0', '0']]
print sol.numIslands(grid)

File: 200_Number_of_Islands/numIslands.cpp
#include <vector>

using namespace std;

class Solution {
	public:
        void removeIsland(vector<vector<char> >& grid, int y, int x) {
            int dx[4] = {1, -1, 0, 0};
            int dy[4] = {0, 0, 1, -1};
            int m = grid.size(), n = grid[0].size();

            grid[x][y] = '0';
            for(int i = 0; i < 4; i++) {
                int nextX = x + dx[i];
                int nextY = y + dy[i];
                if(nextX >= 0 && nextX < n &&
                   nextY >= 0 && nextY < m &&
                   grid[nextX][nextY] == '1') {
                    removeIsland(grid, nextX, nextY);
                }
            }
        }
    	int numIslands(vector<vector<char> >& grid) {
            int res = 0;
            for(int y = 0; y < grid.size(); y++) {
                for(int x = 0; x < grid[0].size(); x++) {
                    if(grid[y][x] == '1') {
                        res++;
                        removeIsland(grid, y, x);
                    }
                }
            }
            return res;
    	}
};
